### ISSUE TITLE
SPM support not working on M1 Macs

### ISSUE THREAD

**Issue by shaps80** [onfido/onfido-ios-sdk#216]
### What was the expected behaviour?

Including the dependency via Swift Package Manager should work as expected

### What happened instead?

The package resolves correctly, however symbols and types cannot be found by the compiler, resulting in code that cannot be compiled.

### Version info:

- Onfido SDK (Debug/Release): Release
- iOS: Any
- Xcode: 12.3
- Device/Simulator: Both

### Integration configuration:

You sent me a test project, same issue.

### Steps to reproduce:

Using your test project as provided on #182, simply open it on an M1 and try to compile.

**Comment by anuragajwani** [comment]
Awesome thanks for opening this separate issue @shaps80. The upcoming release does not include fixes for the M1 Mac.

**Comment by shaps80** [comment]
any progress on this?

**Comment by reddavis** [comment]
+1. All our new iOS devs are receiving M1 macs and this is rather problematic.

**Comment by shaps80** [comment]
Ditto, I'm literally growing my team right now and we're all going to be on M1's.

Side note, thank you for making this a priority and working to get this out. Its really appreciated and the communication has been great 👍

**Comment by anuragajwani** [comment]
sorry for the delay unfortunately I am not able to change my work Mac as there is no M1 Mac option with 32GB Ram yet. Going through the process of acquiring a testing device and awaiting from the business to respond.

Out of personal interest are you all acquiring 16GB ram (It's the max option on MacBook Pro and Mac mini's with M1 chips)? I am currently on 16GB and my Mac really struggles to keep up with my demand. So I am interested to understand why 16GB is sufficient for you all.

**Comment by shaps80** [comment]
I have an older 2018 Intel Mac mini with 32Gb, but it rarely required using the more than 16 in my experience.

Are you using a lot of VMs for example? That's a prime example of where you'd have excessive memory requirements. Otherwise it really shouldn't be an issue IMHO.

CPU/Memory architecture isn't really my background so I'm not sure to be fair. But for iOS development, UI design, video editing, etc..., I've never really found I needed more than 16GB.

The reason I have the Intel on 32GB was because I _do_ sometimes use a little more during music production. Instruments tend to use up a lot more RAM so the more you use, the more you need. I also now use it as my 'server' as the M1 has actually handled that perfectly as well ¯\_(ツ)_/¯ 

Sorry, not sure if that's helpful for you or not.

---

Is there something we can do to help test @anuragajwani ? I'm really keen to have this working correctly. Right now I've just included your xcframework directly in the project but obviously this is now no longer tracking your latest versions and means I also have to include it in our own repo for the moment. I'd be happy to help out if there's an option here? Let me know.

**Comment by anuragajwani** [comment]
@shaps80 sorry if my questions will be basic in that case. I understand that the XCFramework is working when manually imported. However SPM or Cocoapods is not working correctly. Whats the error message or unexpected behaviour you are encoutering?

**Comment by anuragajwani** [comment]
On a personal note I only use my work laptop (16GB) for iOS development. It is super slow. I am not talking about build times yet I am just talking about wake time, entering my password then loading the screen. Changing tabs is also slow and yeah I have cleared browser tabs and restarted to see whether that helps. Nothing.

It also could be because we have a lot of security software installed which really consumes a good chunk of the memory.

**Comment by shaps80** [comment]
Yeah security software, antivirus etc definitely have have a strong negative impact. I've had that over the years myself. Pity. Also Chrome, yuck!

As for the XCFramework, when I import it via SPM, certain symbols and types can't be resolved. I honestly don't know why its happening. And without being able to try and compile it myself, I can't be more specific I'm afraid.

**Comment by shaps80** [comment]
MY guess if you just need to compile it on an M1 so that architectures are setup correctly?

**Comment by shaps80** [comment]
Note: Just to clarify, I meant to say I have directly included the FAT framework, NOT the XCFramework for the same reasons.

**Comment by anuragajwani** [comment]
> As for the XCFramework, when I import it via SPM, certain symbols and types can't be resolved

That helps. However I though Xcode would be smarter. I assumed that M1 Macs would be running on arm64 architecture (same as iPhones) however that sounds not be the case. With XCFrameworks the frameworks are organised in platform + architecture basis.

Our XCFRamework includes two frameworks:
1. iOS-arm64_armv7 (armv7 being the old CPU architecture that we need to support for legacy apps customers)
2. iOS-i386_x86_64-simulator (simulators on MacOS until M1 Macs)

I assume that M1 Macs runs on arm64 architecture and thus the symbols should be under the arm64 directory.

Quick question do you get symbols not found for both simulators and devices? Or just one of them?

**Comment by shaps80** [comment]
> > As for the XCFramework, when I import it via SPM, certain symbols and types can't be resolved
> 
> That helps. However I though Xcode would be smarter. I assumed that M1 Macs would be running on arm64 architecture (same as iPhones) however that sounds not be the case. With XCFrameworks the frameworks are organised in platform + architecture basis.

Yeah so they are ARM64, which is why with the FAT framework, I had to strip the intel arcs to make it work.

> Quick question do you get symbols not found for both simulators and devices? Or just one of them?

Good question, I'll check and get back to you today.

**Comment by shaps80** [comment]
Yeah, so device works actually. I'm not sure if I checked that before sorry.

That might unblock me for now though as long as I can ship a build with it.

Not sure why Simulators are problematic though. Intercom recently updated their SDK similarly (i.e. XCF via SPM) and that works perfectly. So there's obviously _some_ requirement but I'm not sure what it is. Maybe just has to be built on M1?

**Comment by anuragajwani** [comment]
@shaps80 as a quick workaround you can use the framework contained within the XCFramework maybe. If you download and open the XCFramework and navigate to the arm64_armv7 build only you won't need to strip anything from that.

![Screenshot 2021-02-12 at 14 36 57](https://user-images.githubusercontent.com/7071488/107781612-fbcfd100-6d3f-11eb-94a0-f65a27f8a676.png)
![Screenshot 2021-02-12 at 14 37 22](https://user-images.githubusercontent.com/7071488/107781632-012d1b80-6d40-11eb-974d-290d6fdecb32.png)

Let me know if this works please 😄

**Comment by shaps80** [comment]
Ok, so that does give me a potential hint towards the issue. When I try that, it then tells me that the simulator version is missing. So I'm guessing we need an XCFramework that includes simulator architectures for ARM?

**Comment by anuragajwani** [comment]
yeah I started suspecting that Xcode is using the simulator build for intel based Macs on M1 Macs maybe just because of the `-simulator` naming on the directory. I think on our side we need to add arm64 to the simulator build thus the directory would be named something like `ios-i386_x86_64_arm64-simulator`. 

Of course I'm doing guess work without access to an M1 Mac at this point.

Still waiting confirmation from the finance team for the M1 Mac 😅

**Comment by shaps80** [comment]
Yeah that's my thought as well. I'll try and investigate further and let you know if I find anything.

**Comment by shaps80** [comment]
![Screenshot 2021-02-12 at 16 00 42](https://user-images.githubusercontent.com/251310/107791188-8833c100-6d4b-11eb-8c9e-d16ef2e4cfcd.png)
Intercom works and is doing as you've suggested 👍

**Comment by anuragajwani** [comment]
well that was a lucky guess 😂. I am going to try get some time to make a build this way. If so would you be able to test it out for us please?

**Comment by shaps80** [comment]
yeah definitely!

**Comment by shaps80** [comment]
I'll keep an eye on this issue for an update thanks :)

**Comment by anuragajwani** [comment]
I finally got an approval from finance. Nevertheless I'll make that build most likely on Monday and share it with you.

Thanks for the helping us out 😄 this should really help us to get the fix on the actual release once I get the M1 Mac

**Comment by anuragajwani** [comment]
hey @shaps80 I gave the build a shot this evening. It looks like whilst the xcodebuild command successfully builds the Onfido SDK with the iphonesdk set to `iphonesimulator` and the architecture set to `arm64` building from an Intel based Mac it does not produce the binary 🤔 . 

On the other hand the good news: Onfido has approved my purchase of an M1 Mac and should order it soon for me.

**Comment by shaps80** [comment]
That's interesting. I'd assumed there was a way to produce a binary for Apple Silicon from an Intel Mac, but perhaps that's not currently the case? Glad to hear you're getting one and this will soon be resolved! Appreciate the work you've been putting into this.

**Comment by anuragajwani** [comment]
even setting ARCHS build var to `i386 x86_64 arm64` only product the binary with `i386` and `x64_64`. 

```
xcodebuild build ... -derivedDataPath derived_data -sdk iphonesimulator -configuration Debug  ARCHS="i386 x86_64 arm64" | xcpretty
```

=> 

```
derived_data/Build/Products/Debug-iphonesimulator/Onfido.framework/Onfido (for architecture x86_64):	Mach-O 64-bit dynamically linked shared library x86_64
derived_data/Build/Products/Debug-iphonesimulator/Onfido.framework/Onfido (for architecture i386):	Mach-O dynamically linked shared library i386
```

Thanks for the patience too though and sorry for the inconvenience. Hopefully I can get that M1 Mac asap.

**Comment by anuragajwani** [comment]
Quick update: I haven't received yet the M1 Mac mini from Apple yet. Still stuck on prepping order unfortunately.

**Comment by dimat** [comment]
I believe you should be able to cross-compile for arm64 even on Intel mac. Just double check the architectures that you have either in build settings, or if you have a script that joins multiple archs into a fat library.

At the moment you have only following files for simulator covering i386 and x86_64 architectures:
```
i386-apple-ios-simulator.swiftdoc
i386-apple-ios-simulator.swiftinterface
i386-apple-ios-simulator.swiftmodule
x86_64-apple-ios-simulator.swiftdoc
x86_64-apple-ios-simulator.swiftinterface
x86_64-apple-ios-simulator.swiftmodule
```

You need to add one more `arm64` for simulator (you already have it for device).

**Comment by anuragajwani** [comment]
@dimat that is the plan. We did this with fat frameworks before XCFrameworks became available. The place where I feel most uncertain is using iphoneos sdk vs iphonesimulator sdk for arm64 targeted to simulators. `iphoneos` sdk seems to work just fine on M1 Mac based on some of the responses here yet when using the XCFramework Xcode selects simulator. I assume the `iphoneos` is installed then on the M1 based Mac which is not exactly surprising given the M1 Mac can actually run native iOS apps.

Still without an M1 Mac I can't thoroughly test the Onfido SDK and decide which is the appropriate build profile to choose here. We might even need to make some code changes as we treat `iphoneos` always having a camera and simulator not having them (this is just the first example that pops in my head).

**Comment by dimat** [comment]
Although it is possible to run iOS apps on the mac, I think, there is difference whether you compile with arm64 with iphoneos sdk or iphonesimulator.

You can see that architecture and SDK are two different parameters in xcodebuild (I took this snippet from an example):
```
xcodebuild -target "${FRAMEWORK_NAME}" -configuration Release -arch arm64 -arch armv7 -arch armv7s only_active_arch=no defines_module=yes -sdk "iphoneos"
xcodebuild -target "${FRAMEWORK_NAME}" -configuration Release -arch x86_64 -arch i386 only_active_arch=no defines_module=yes -sdk "iphonesimulator"
 ```

If you have something similar in your codebase, you might just need to include `-arch arm64` where you have `-sdk "iphonesimulator"`.

If it helps, I am happy to help with testing.

**Comment by anuragajwani** [comment]
@dimat I am not sure if I am completely following you message. However I have the problem is mainly that I can't build the Onfido SDK with -arch arm64 AND -sdk iphonesimulator. I am questioning whether I even need the iphonesimulator as the sdk instead of iphoneos. However we assume certain things when we compile using iphoneos sdk (associated build profile). We assume it will be used in an actual device with camera in it.

So our current build script build arm64 iphoneos specifically with that in mind and now that might not be the case as you can have a M1 Mac mini with no camera. Also now using iphoneos and arm64 we'd want a separate a build to the current arm64 build that takes into account that simulator like logic.

So even if today I combine arm64 slice into the simulator build there is no guarantee that it will work as our integrators are used to simulator build. That is being able to skip certain on device validations for testing purposes and seeing a black screen when no camera is available.

I can easily combine the framework in the blind however without thorough testing I can' guarantee features will work expected. Hence why I need an M1 Mac to carry out that testing.

**Comment by shaps80** [comment]
Any ideas how much longer until you get one? We've shifted some stuff to SwiftUI and its proving difficult since without Simulator support, we can't run SwiftUI previews either 🤣

**Comment by anuragajwani** [comment]
@shaps80 we haven't got onto SwiftUI. I think that is a separate issue which I have raised with the rest of the team.

On the M1 Mac mini still haven't received it. It's taking way long time than usual. I am going to check again with IT department see what's the status on that.

**Comment by shaps80** [comment]
Ok sure. The SwiftUI stuff doesn't affect you, was just an interesting observation :)

**Comment by anuragajwani** [comment]
Finally received the M1 Mac mini. Will set it up today.

**Comment by anuragajwani** [comment]
Hey all quick update:

The reason why Onfido SDK is not building for iphonesimulator arm64 is due to one of our dependencies which is not needed for simulators.

So I tried a few hacks:
- create fat simulator framework with iPhoneos arm64 slice using lipo  -> Xcode throws incompatibility issue on consumption
- create simulator framework using -create-xcframework with iphoenos arm64 -> xcodebuild throws incompatibility error (iphonesimulator and iphoneos slices can't mix)

The other option where we set the arches build var to "arm64 x86_64" was producing only the x86_64 slice when targeting iphonesimulator sdk. I think this is related to the dependency that is not compatible with iphonesimulator arm64.

We are contemplating to remove the dependency from simulator builds but that will require some code changes.

Just wanted to post to assure you we are working on this right now.

Sorry for the inconveniences.

**Comment by seewah** [comment]
Hi just an update on this. We have been doing more work on this. Unfortunately the work required to refactor our SDK to get it working on M1 macs turns out to be more complicated than expected. We will have to park it for now and pick it up again w/c 22nd March. Hopefully we will have a solution by the end fo March. Apologies for the inconvenience.

**Comment by shaps80** [comment]
That's fine, EO March would be amazing. Honestly appreciate all the effort you guys are putting in to get this fixed for us!

**Comment by shaps80** [comment]
@seewah / @anuragajwani just following up, still on track for a soon-ish build?

**Comment by seewah** [comment]
Hi it is in our current sprint, which started today. So we will see how we get on. We will post an update once we have a solution, and then we can provide a more accurate ETA? thanks

**Comment by anuragajwani** [comment]
Hey all we currently have 2 working solutions which are being reviewed and tested. This will most likely released for 21.3.0. 21.2.0 is coming up soon and won't have the fix. ETA for 21.3.0 is up to 3 weeks.

**Comment by shaps80** [comment]
3 weeks? Yikes, any chance I can get an early release (even a potentially buggier snapshot?)

Appreciate the effort you’re all putting in btw, thanks again!

**Comment by anuragajwani** [comment]
@shaps80 trying to kick off the release asap. I am trying to make a release out of cycle as we just completed 21.2.0. If I don't make it by next week I'll share a dirty version of the SDK that supports M1 Mac in this thread.

**Comment by shaps80** [comment]
Super appreciated!

**Comment by anuragajwani** [comment]
@shaps80 we are near to finalise the release 21.3.0 which supports M1 Macs. Until its not merged you can use commit `18b9e4feb8fe1230a02c14cb8b6b7c40ce2ac78f` with SPM to refer to this version.

**Comment by shaps80** [comment]
Amazing!! We're in the middle of our own release, but in the next couple days I'll try and give this a shot! Amazing work and thank you so much for all the effort you all have put in to make this happen! Super appreciated!

**Comment by anuragajwani** [comment]
no worries 😄 thanks for the patience. We have now finalised the release so you can test it out using SPM + 21.3.0 tag.

**Comment by anuragajwani** [comment]
Please let us know if you encounter any issues or if we can close the ticket. Thanks!

**Comment by shaps80** [comment]
@anuragajwani works like a charm!!! Amazing work and thank you so much for getting this done. Smooth sailing ahead :)
